core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java

```diff
@@ -0,0 +1,239 @@
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.endpoint.FileWatcherRunnable.FileWatchEvent;
+
+/**
+ * Wraps a {@link WatchService} and allows paths to be registered.
+ */
+final class FileWatcherRegistry implements AutoCloseable {
+
+    /**
+     * Context responsible for watching a {@link FileSystem}. Contains a reference to
+     * a {@link WatchService} and a {@link Thread} which continuously watches for changes
+     * in registered file paths.
+     */
+    static class FileWatchServiceContext {
+
+        private final RestartableThread restartableThread;
+        private final WatchService watchService;
+        private final Map<FileWatchRegisterKey, FileWatchEvent> currWatchEventMap =
+                new ConcurrentHashMap<>();
+
+        /**
+         * Creates a new {@link FileWatchServiceContext}.
+         * @param name the name of the thread used to watch for changes
+         * @param watchService the {@link WatchService} used to monitor for changes
+         */
+        FileWatchServiceContext(String name, WatchService watchService) {
+            restartableThread = new RestartableThread(name, () ->
+                    new FileWatcherRunnable(watchService, this));
+            this.watchService = watchService;
+        }
+
+        /**
+         * Starts to watch changes for the path corresponding to the {@link FileWatchRegisterKey}.
+         * When changes are detected, the {@code callback} is invoked. If the {@link WatchService}
+         * isn't running yet, this method starts a thread to start watching the {@link FileSystem}.
+         * @param watchRegisterKey key that contains the path to be watched
+         * @param callback function invoked on file change
+         *
+         * @throws IllegalArgumentException if failed to locate file or failed to start watching
+         */
+        void register(FileWatchRegisterKey watchRegisterKey, Runnable callback) {
+            final Path dirPath;
+            try {
+                dirPath = watchRegisterKey.filePath().toRealPath().getParent();
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to locate file " +
+                                                   watchRegisterKey.filePath(), e);
+            }
+
+            final WatchKey watchKey;
+            try {
+                watchKey = dirPath.register(watchService,
+                                            ENTRY_CREATE,
+                                            ENTRY_MODIFY,
+                                            ENTRY_DELETE,
+                                            OVERFLOW);
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to watch file " + watchRegisterKey.filePath(), e);
+            }
+            currWatchEventMap.put(watchRegisterKey, new FileWatchEvent(watchKey, callback, dirPath));
+            if (!currWatchEventMap.isEmpty()) {
```

Comment:
you're right, we probably don't need this check.
(Let me know if you also think the method name should be changed to something more appropriate like `startIfPossible`... I think as-is is fine though since a single thread can't start multiple times)